File: Suho/무인도 여행/무인도 여행.py
from collections import deque
def solution(maps):
    #bfs문제
    answer = []
    map_row = len(maps)
    map_col = len(maps[0])
    visited=[[0]*map_col for _ in range(map_row)]
    dx=[0,0,-1,1]
    dy=[-1,1,0,0]
    
    def bfs(x,y):
        q = deque([(x, y)])
        visited[x][y] = 1
        total = int(maps[x][y])
        while q:
            x, y = q.popleft()
            for i in range(4):
                nx = x+ dx[i]
                ny = y+ dy[i]
                if 0<=nx< map_row and 0<=ny< map_col:
                    if not visited[nx][ny]and maps[nx][ny] !='X':
                        visited[nx][ny]=1
                        q.append((nx,ny))
                        total+=int(maps[nx][ny])
        return total
    for i in range(map_row):
        for j in range(map_col):
            if not visited[i][j] and maps[i][j] !='X':
                total=bfs(i,j)
                answer.append(total)
   #만약 지낼 수 있는 무인도가 없다면 
    if not answer:
        return [-1]
    #오름차순배열
    return sorted(answer)

    

File: Seojin/무인도 여행/무인도 여행.java
import java.util.*;
class Solution {
    public int[] solution(String[] maps) {
        int[] answer = {};
        
        boolean[][] visited = new boolean[maps.length][maps[0].length()];
        
        ArrayList<Integer> iland = new ArrayList<>();
        int[][] dir = {{1,0},{-1,0},{0,1},{0,-1}};
        for (int i=0;i<maps.length;i++){
            for (int j=0;j<maps[0].length();j++){
                if (!visited[i][j] && maps[i].charAt(j) != 'X'){
                    int sum = 0;
                    Deque<int[]> d = new ArrayDeque<>();
                    d.offer(new int[]{i,j});
                    visited[i][j] = true;
        
                    while(!d.isEmpty()){
                        int[] cur = d.poll();
                        sum+=maps[cur[0]].charAt(cur[1])-'0';
                        
                        for (int k=0;k<4;k++){
                            int cx = cur[0]+dir[k][0];
                            int cy = cur[1]+dir[k][1];
                            if (cx>=0 && cx<maps.length && cy>=0 && cy<maps[0].length() && !visited[cx][cy] && maps[cx].charAt(cy) != 'X'){
                                visited[cx][cy] = true;
                                d.offer(new int[]{cx,cy});
                            }
                        }
                    }
                    iland.add(sum);
                }
            }
        }
        
        Collections.sort(iland);
        if (iland.size()>0){
            answer = new int[iland.size()];
            for (int i=0;i<iland.size();i++){
                answer[i]=iland.get(i);
            }
        }
        else answer = new int[]{-1};
        
            
        return answer;
    }
}
